backend/src/main/repository/PromotionRepository.ts

```diff
@@ -1,5 +1,121 @@
-import { EntityRepository, Repository } from 'typeorm';
+import { EntityRepository, Repository, SelectQueryBuilder } from 'typeorm';
 import { Promotion } from '../entity/Promotion';
+import { PromotionQueryDTO } from '../validation/PromotionQueryValidation';
 
+/* eslint-disable  @typescript-eslint/no-explicit-any */
+/* eslint-disable  @typescript-eslint/explicit-module-boundary-types */
 @EntityRepository(Promotion)
-export class PromotionRepository extends Repository<Promotion> {}
+export class PromotionRepository extends Repository<Promotion> {
+  /**
+   * Gets all promotions by constructing a query builder.
+   * * Depending on whether promotionQuery has any content, we will need to apply the necessary operations on top of
+   * the queryBuilder in order to support filtering operations
+   * */
+  async getAllPromotions(promotionQuery?: PromotionQueryDTO): Promise<any> {
+    if (
+      promotionQuery &&
+      JSON.stringify(promotionQuery) !== JSON.stringify({})
+    ) {
+      return this.applyQueryOptions(promotionQuery);
+    } else {
+      return this.createQueryBuilder('promotion')
+        .innerJoinAndSelect('promotion.discount', 'discount')
+        .getMany();
+    }
+  }
+
+  /**
+   * Depending on which properties are defined inside promotionQuery, we add those properties into our query for the queryBuilder to execute.
+   */
+  private applyQueryOptions(promotionQuery: PromotionQueryDTO): Promise<any> {
```

Comment:
I thought about this a bit more and was wondering if you also agree the filtering/sorting should be done server-side? I thought that it would be nice to do the queries as close to the DB as possible, but I forgot that that means making a network request each time the user wants to filter/sort. Another thought that I had is that we'd be displaying the promotions within each filter ordered by distance (my assumption), so that would mean we'd have to do processing work on the FE anyways to retrieve lat/lon data. But maybe you have other thoughts on this?

Reply:
yeahh like this is def something I wanted to bring up, I think I like sorting in the front end, that way it can be much faster without the need to make a request each time. (like if we perform a complex search query + filters, and get the results, then the user wants to sort, we will need to execute the complex search query + filters again in addition to the sort)

But I think filtering itself can be a network request, mainly because of what you mentioned about queries as close to the DB as possible. https://stackoverflow.com/questions/52346685/filters-logic-should-be-on-frontend-or-backend#:~:text=This%20depends%20on%20the%20size,end%20after%20getting%20the%20data.

At the same time, I've attached the `rank` from PostgreSQL's full-text search, so you still have the info on which promotions are most relevant to the search query so that the FE can sort by relevance (rank).

Reply:
Let me know what you think about this!

Reply:
@ericrkuo Yeah that makes sense! Also, filtering on BE is nice because it minimizes the packet sizes of our network requests (although this probably won't be a big problem for us since we're not dealing with big data). I think we'll have to do some combination of sorting by several keys (e.g. rank, distance) depending on different situations but that can all be handled by FE so this works

Reply:
Awesome sweet!

Reply:
Yah that makes sense about packet sizes. I'm thinking that in the future we will encounter this problem about packet sizes as we will have quite a lot of data to send back. But I think it should be okay hopefully for now.